### DISCUSSION TITLE
Question about reading singleton collection and collection names

### DISCUSSION THREAD

**Discussion by sagerio** [directus/directus#5111] [Archive]
Hi ho,

some minor questions came up while developing with the last rc59:

- When a collection is defined as singleton, it has of course only one row, normally with ID = 1.
Is it correct that one can read this row either with `sdk.items("collectioname").readOne(1)` or with `sdk.items("collectioname").readMany({ single: true })` ?

- Taking the example from above `sdk.items("collectioname").readOne(1)`, why it comes that this works but taking the collection name from a variable like `sdk.items(environment.collections.collectionname).readOne(1)` doesn't? (Result is typeof `PartialItem<unknown>` instead of `PartialItem<collectionname>`)

Thanks a lot!

**Comment by rijkvanzanten** [comment]
> When a collection is defined as singleton, it has of course only one row, normally with ID = 1.
Is it correct that one can read this row either with sdk.items("collectioname").readOne(1) or with sdk.items("collectioname").readMany({ single: true }) ?

There's a dedicated `.singleton("collectionname").read();` 🙂 

@WoLfulus can we document this a little more clearly in the SDK docs?

**Reply by rijkvanzanten** [reply]
Not sure about the second question.. @WoLfulus thoughts?

**Reply by sagerio** [reply]
> There's a dedicated .singleton("collectionname").read(); 🙂

Indeed, that's great, thanks! 🚀

**Reply by sagerio** [reply]
The second thing: It works when using a 'normal' string variable:

`const myvar = "collectionName";
sdk.items(myvar).readOne(1);`

Only with the environment thing from Angular it doesn't work, at least the Intellisense shows "unknown"...

**Reply by rijkvanzanten** [reply]
I suppose that's because angular's value might be dynamic which in turn means typescript can't do static analysis on it? Not sure though 🙂

**Reply by WoLfulus** [reply]
The first parameter has no specific type (like string for example), it's a generic type so we can extract it at build type and make a lookup in the type mapping (passed on constructor). This allows us to make sure we return the proper collection type thus avoiding `collection<Type>("col")` all the time. This means that whatever type you pass to the first parameter will be used as the lookup key, and since you don`t have one in that case it will show as unknown.

You can still cast the first parameter to something else.

Does this fail to compile though? Or just not getting proper auto completion?

**Reply by sagerio** [reply]
> Does this fail to compile though? Or just not getting proper auto completion?

Seems like compiling succeeds, it's only the auto completion that says 'unknown' and is therefore empty...

**Reply by WoLfulus** [reply]
Yeah, that makes sense since it's a variable and you won't know at compile time the value in it. You can try casting it to a string to force a type resolution though.

`collection(myvar as "collectionname")`

**Reply by sagerio** [reply]
OK but then it would be easier to write the collectionname alone `collection("collectionname")` :relaxed:
The goal was to avoid repeating the names as string every time and instead take it from the environent.
But as said, that's no big deal...

**Reply by WoLfulus** [reply]
I would need a more complete example so I can suggest something better in your case. I'm not sure what you mean by repeating in this case. Multiple calls to collections? Multiple different collections?

I can look into providing a reverse lookup function for type => name if that's the case, but as I said I'd need a more complete example of what you're doing

**Reply by rijkvanzanten** [reply]
> and instead take it from the environent.

I'm not familiar with Angular, but there's a good chance you can set the type of this environment variable somewhere more appropriate!

**Reply by sagerio** [reply]
Well, in Angular in have a singleton service 'directusService', and in different components I use the Directus instance from this service and make some calls, like `this.directusService.Directus.items("users").readMany(.....)`
So I have to repeat this collection name in every component.

**Reply by rijkvanzanten** [reply]
You could also consider making it something more like:

```
this.users.readMany(...)
```

by relying on the 

```js
const users = directus.items('users');
```

approach 🤔 that should give you the "global typing" in this case

**Reply by sagerio** [reply]
You mean `const users = directus.items('users');` in the singleton service and using this in the components?
That's a very good hint, would be much easier... Shame on me :grin:
I'll try this approach...
Thank you both so much!

**Reply by rijkvanzanten** [reply]
Yeah exactly! That way you can just do `this.directusService.users.readMany()` (or whatever naming scheme you come up with) and still rely on all the typing goodness without having to share an additional variable w/ type info
